torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -2261,20 +2261,28 @@ size_t ONNXAssignOutputShape(
   } else if (THPUtils_checkString(output_obj)) {
     // Ignore string, since they are not supported as output in ONNX.
   } else if (PyNone_Check(output_obj)) {
-    // TODO: Currently there's no one thing to do here that works for
-    // both tracing and scripting.
-    // If we don't increment outputs_index here, then scripting fails
-    // for
-    // `python test/onnx/test_pytorch_onnx_no_runtime.py`.
-    // If we do increment it, then tracing fails for
-    // `python test/onnx/test_pytorch_onnx_onnxruntime.py
-    //  TestONNXRuntime.test_tuple_with_none_outputs`.
+    // Tracing:
+    //    Ignore None, since it is not captured in IR graph as output.
+    // Scripting:
+    //    Ignore None, if observing a fixed `None` node in IR graph. Because it
+    //    is meaningless to include it as graph output as it carries no
+    //    data/information. Plus that static `None` is not supported in ONNX IR.
+    //    Otherwise, the output should have type `Optional`, and should be
+    //    converted to ONNX `Optional`.
+
+    // More context:
     // Cause: in tracing we flatten the outputs in ONNXTracedModule.forward
-    // in torch/jit/_trace.py while tracing. This means the output has None
-    // objects omitted. But then the outputs passed in here are un-flattened,
-    // which means they contain None objects.
-    // Ideally we'd remove this difference.
-    outputs_index += static_cast<size_t>(is_script);
+    // in torch/jit/_trace.py while tracing. This means the traced IR graph
+    // has None outputs omitted.
+    // But then the outputs passed in here are un-flattened, which means they
+    // contain None objects. Ideally we'd remove this difference.
+    if (is_script && outputs_index < graph->outputs().size()) {
+      if (graph->outputs().at(outputs_index)->node()->mustBeNone()) {
+        graph->eraseOutput(outputs_index);
+      } else {
```

Comment:
Q: I am reading the related Optional type code, wondering is this else branch for Optional type output?

Reply:
Yes it is for optional type output. The comment above explains when and why this happens.